### PR TITLE
add issue.py - create a new github issue

### PR DESCRIPTION
**Pull request by amuritna**
for:
- https://codein.withgoogle.com/dashboard/task-instances/5046772335378432/
based on:
- https://github.com/apertium/phenny/issues/433

### CODE REVIEW

File: modules/issue.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+"""
+issue.py - create a new GitHub issue
+author: amuritna
+"""
+
+import requests
```
Comment (sushain97) [inline modules/issue.py]: Please use the `web` module when possible.

File: modules/issue.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import json
+
+oauth_token = 'PUT OAUTH TOKEN HERE'
```
Comment (sushain97) [inline modules/issue.py]: This should come from the `phenny` config, not the module.

File: modules/issue.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    """ .issue <repository> <title> - create a new GitHub issue """    
+ 
+    gh_check = requests.head(gh_uri)
+    if gh_check.status_code != 200:
+        return phenny.reply('Sorry, GitHub is down.')
+
+    # check input validity
+    try:
+        if not input.group(1):
+            return phenny.reply('For which repository do I make this issue?')
```
Comment (sushain97) [inline modules/issue.py]: Instead, just print out the usage. Similar comment below.

File: modules/issue.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+        content = input.group(1).strip()
+
```
Comment (sushain97) [inline modules/issue.py]: Unnecessary newline.

File: modules/issue.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        repo = ghpath[1]
+
+        if owner not in allowed_owners:
+            return phenny.reply('Sorry, you don\'t have permission to create an issue there.')
+
+        title = " ".join(content.split()[1:]).strip()
+        if len(title) < 1:
+            return phenny.reply('What is the title of the issue?')
+
+    except:
```
Comment (sushain97) [inline modules/issue.py]: Never use bare excepts in Python.

File: modules/issue.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        
+        # check whether likely in an owner/repository combo format
+        if len(ghpath) != 2:
+            return phenny.reply('Sorry, that doesn\'t look like a valid repository path.')
+
+        owner = ghpath[0]
+        repo = ghpath[1]
+
+        if owner not in allowed_owners:
+            return phenny.reply('Sorry, you don\'t have permission to create an issue there.')
```
Comment (sushain97) [inline modules/issue.py]: This isn't really the correct error message. It's more like begiak can't do it.

**Comment by amuritna** [comment]
Thank you very much for the feedback! Right now I'm working on replacing requests with the web module.

File: modules/issue.py (hunk 1 above)
Comment (Androbin) [reply modules/issue.py]: We can control caching and timeouts more easily.

File: modules/issue.py [hunk 7]
```diff
@@ -2,48 +2,51 @@
...
-        title = " ".join(content.split()[1:]).strip()
-        if len(title) < 1:
-            return phenny.reply('What is the title of the issue?')
+    	title = " ".join(content.split()[1:]).strip()
+    	if len(title) < 1:
+        	return phenny.reply(invalidmsg)
 
-    except:
-        return phenny.reply('For which repository do I make this issue?')
+    except SyntaxError:
```
Comment (Androbin) [inline modules/issue.py]: Where would `SyntaxError` be raised?

Comment (amuritna) [reply modules/issue.py]: Using the method `.group(1)` on `input` raised a SyntaxError whenever I locally tested it with the command `.issue` (without any parameters). Do you have any suggestion on how to deal with this better?

**Comment by amuritna** [comment]
Hm, I've searched through StackOverflow and the webpy documentation, and it's looking like the web module was designed with a way to send POST requests to external URLs (which is what we need to do)? [Their own docs recommended](http://web2py.com/books/default/chapter/34/14/other-recipes#Fetching-an-external-URL) urllib and fetch. Other modules which need to do something similar - e.g. search.py - also use requests. I am by no means a webpy expert, though, so I really can use some help right now.

**Comment by Androbin** [comment]
@amuritna It's not `webpy`, the framework. It's `web.py`, our own module.

File: modules/issue.py (hunk 7 above)
Comment (Androbin) [reply modules/issue.py]: I couldn't reproduce that error. I removed the `try-except` block and ran the command `.issue` without parameters. No error was raised.

Comment (Androbin) [reply modules/issue.py]: Also, `SyntaxError` is about an issue with the syntax of your code, not the input.

**Comment by amuritna** [comment]
That was very embarrassing. 

I'll go work on it.

File: modules/issue.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
+"""
+issue.py - create a new GitHub issue
+author: amuritna
+"""
+#following code might or might not be uncommented,
+#depending on how your Python setup
+#import sys
+#sys.path.insert(0, '..')
```
Comment (sushain97) [inline modules/issue.py]: wat? please remove this. this code should be running via phenny or tests

File: modules/issue.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+def issue(phenny, input):
+	""" .issue <owner>/<repository> <title> - create a new GitHub issue """    
+ 
+	if not is_up(gh_uri):
+		return phenny.reply('Sorry, GitHub is down.')
+	
+	# check if GitHub auth token is available in default.py
+	try:
+		if phenny.config.gh_oauth_token:
```
Comment (sushain97) [inline modules/issue.py]: Why do you need this if you're already using the attribute error below?

File: modules/issue.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+			oauth_token = phenny.config.gh_oauth_token
+	except AttributeError:
+		return phenny.reply('GitHub authentication token needs to first be set in the configuration file (default.py)')
```
Comment (sushain97) [inline modules/issue.py]: Can we just disable this module unless the config entry is set? cc @Androbin who I think I asked to do something similar earlier.

File: modules/issue.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+		if owner not in allowed_owners:
+			phenny.reply('Begiak cannot create an issue there.')
+			return isinvalid
+
+		title = " ".join(content.split()[1:]).strip()
+		if len(title) < 1:
+			phenny.reply(invalidmsg)
+			return isinvalid
+		
+	except SyntaxError:
```
Comment (sushain97) [inline modules/issue.py]: Why is this here? I think it should be entirely removed.

File: modules/issue.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+	req_headers = {'Authorization': 'token ' + oauth_token}
+	req_body = json.dumps({
+		"title": title,
+		"body": default_desc
+	})
+
+	req_str = post(req_target, req_body, req_headers)
+	req_json = json.loads(req_str)
+
+	return phenny.reply('Issue created. You can add a description at ' + req_json["html_url"])
```
Comment (sushain97) [inline modules/issue.py]: Let's stick with one type of quote. Single or double, not both.

File: modules/issue.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+		if len(title) < 1:
+			phenny.reply(invalidmsg)
+			return isinvalid
+		
+	except SyntaxError:
+		phenny.reply(invalidmsg)
+		return isinvalid
+
+	# build and post HTTP request
+	req_target = gh_uri + '/repos/' + owner + '/' + repo + '/issues'
```
Comment (sushain97) [inline modules/issue.py]: Use format-strings please.

File: modules/test/test_issue.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+
+class TestIssue(unittest.TestCase):
+    def setUp(self):
+        self.phenny = MagicMock()
+        self.input = MagicMock()
+        
+    def test_issue(self):
+        test = ['.issue', 'octocat/Hello-World Create an illegal issue.']
+        self.assertTrue(issue.issue(test) == 'invalid input')
+        test = ['.issue', 'apertium/hypothetical-long-named-repository-apertium-hopefully-would-never-create Create an issue on a nonexistent repo.']
```
Comment (sushain97) [inline modules/test/test_issue.py]: These should be different tests (i.e. functions)

File: modules/issue.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+	
+	# check if GitHub auth token is available in default.py
+	try:
+		if phenny.config.gh_oauth_token:
+			oauth_token = phenny.config.gh_oauth_token
+	except AttributeError:
+		return phenny.reply('GitHub authentication token needs to first be set in the configuration file (default.py)')
+		
+	# check input validity
+	isinvalid = 'invalid input' # for unit tests
```
Comment (sushain97) [inline modules/issue.py]: This isn't how unit testing should be done. You're testing the wrong thing! Instead, test that phenny replies with `invalidmsg`. See how other module tests do this.

File: modules/issue.py (hunk 10 above)
Comment (Androbin) [reply modules/issue.py]: I recommend adding `'gh_oauth_token': None` to `defaults` in the `phenny` file.
Either you put `if phenny.config.gh_oauth_token:` where appropriate or we change `module_control` to accept `setup` returning `False` to signal self-disabling.

**Comment by sushain97** [comment]
Regarding your first issue, I would check what `web.py` does for that function. Maybe you can change it in a backwards compatible way.

Re second issue, yes. Monkey patching is the right way to check for most requests. Maybe @Androbin knows another module that monkey patches in begiak tests.

**Comment by Androbin** [comment]
```
$ grep -l '@patch' **/test_*.py
modules/test/test_clock.py
modules/test/test_head.py
modules/test/test_search.py
test/test_bot.py
test/test_irc.py
```

**Comment by amuritna** [comment]
I've actually looked into the web.py file, but I didn't really understand it and couldn't place what happened with the return function ... I'll try harder, I guess! Thanks a lot.

**Comment by sushain97** [comment]
Which part are you having trouble understanding? Try stepping through it with a debugger.

**Comment by amuritna** [comment]
Thankfully figured it out with debugging and 'trying harder'. My mistake was not looking deeper into what requests's `raise_for_status()` does (apparently pulling out an error -- makes sense, now). Well, now I'm a smidgen better of a programmer than I was yesterday.

**Comment by sushain97** [comment]
Good to hear you found it! Hmm, I still think that the old way is better since the exception has more information than just an error code.

**Comment by amuritna** [comment]
> I still think that the old way is better
I agree! I've changed it. Should have thought of just trying to catch the error right from the start.

Also, I've attempted some ... monkey patching? I've been trying to read up on it but I still don't feel like I have any idea what I'm doing.

File: modules/test/test_issue.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    
+    def setUp(self):
+        self.phenny = MagicMock()
+        self.input = MagicMock()
+        
+    @patch('modules.issue.post')
+    def testSuccess(self, mock_post):
+        mock_response = MagicMock()
+        mock_response.json.return_value = {
+        'html_url': 'https://github.com/test/test'
```
Comment (sushain97) [inline modules/test/test_issue.py]: Indentation is off here.

File: modules/test/test_issue.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+from modules import issue
+
+class TestIssue(unittest.TestCase):
+    
+    def setUp(self):
+        self.phenny = MagicMock()
+        self.input = MagicMock()
+        
+    @patch('modules.issue.post')
+    def testSuccess(self, mock_post):
```
Comment (sushain97) [inline modules/test/test_issue.py]: These should be snake_case per PEP8. i.e. `test_success`. Please also modify the functions below.

File: modules/test/test_issue.py [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+        'html_url': 'https://github.com/test/test'
+    }
+    mock_post.return_value = mock_response
+    self.input.group.return_value = 'test'
+    issue.issue(self.phenny, self.input)
+    self.phenny.reply.assert_called_with('Issue created. You can add a description at https://github.com/test/test')
+             
+    def testIllegal(self):
+        test = ['.issue', 'octocat/Hello-World Create an illegal issue.']
+        self.assertTrue(issue.issue(test) == 'Begiak cannot create an issue there.')
```
Comment (sushain97) [inline modules/test/test_issue.py]: Rather than asserting that the right value is returned, assert on the `reply` call.

File: modules/test/test_issue.py [hunk 19]
```diff
@@ -0,0 +1,28 @@
+import unittest
+from mock import MagicMock
+from modules import issue
+
+class TestIssue(unittest.TestCase):
+    
+    def setUp(self):
+        self.phenny = MagicMock()
+        self.input = MagicMock()
```
Comment (sushain97) [inline modules/test/test_issue.py]: I think you can just use `lambda x: ...` for this. See other module tests.

File: modules/issue.py [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+
+		title = " ".join(content.split()[1:]).strip()
+		if len(title) < 1:
+			return phenny.reply(invalidmsg)
+		
+	except SyntaxError:
+		return phenny.reply(invalidmsg)
+
+	# build and post HTTP request
+	req target = ('%u/repos/%o/%r/issues' % (gh_uri, owner, repo))
```
Comment (sushain97) [inline modules/issue.py]: These extra parents are unnecessary. Also, I would call this `api_url` or something.

File: modules/issue.py [hunk 21]
```diff
@@ -0,0 +1,71 @@
+"""
```
Comment (sushain97) [inline modules/issue.py]: This entire file uses tabs. I believe all the other code in this repo uses 4 spaces?

File: modules/test/test_issue.py (hunk 19 above)
Comment (amuritna) [reply modules/test/test_issue.py]: I'm unsure where? setUp? Output assertion? From my research all the other established setUp the same way. For the assertions it's for input processing I think, is that what you have in mind?

Comment (sushain97) [reply modules/test/test_issue.py]: Er, I meant in the actual mocks, not here. e.g. https://github.com/apertium/phenny/blob/master/modules/test/test_apertium_wiki.py#L27

Comment (amuritna) [reply modules/test/test_issue.py]: Thank you :)

**Comment by amuritna** [comment]
Is this better?

File: modules/issue.py (hunk 21 above)
Comment (sushain97) [reply modules/issue.py]: This isn't fixed.

File: modules/issue.py (hunk 11 above)
Comment (sushain97) [reply modules/issue.py]: This isn't fixed either.

File: modules/test/test_issue.py [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+        
+    @patch('modules.issue.post')
+    def test_success(self, mock_post):
+        mock_response = MagicMock()
+        mock_response.json.return_value = {
+            'html_url': 'https://github.com/test/test'
+        }
+        mock_post.return_value = mock_response
+        self.input.group = lambda x: ['.issue' 'test/test Create a test issue.'][x]
+        issue.issue(self.phenny, self.input)
```
Comment (sushain97) [inline modules/test/test_issue.py]: You should also make sure that the correct request is sent out here.

Comment (amuritna) [reply modules/test/test_issue.py]: Thanks for all the help! I tried, but I don't know whether I did it correctly. I looked at the other modules' tests and copied whatever seemed missing from issue's, but I doubt that's the correct way of doing things. I'm not sure what to look for when running the tests, or when searching it up.

Comment (sushain97) [reply modules/test/test_issue.py]: You're verifying that `phenny.reply` is being called with the right value. All that remains is to assert that `modules.issue.post` is being called with the right value as well.

File: modules/issue.py (hunk 11 above)
Comment (sushain97) [reply modules/issue.py]: It shouldn't be throwing a syntax error. See https://github.com/apertium/phenny/pull/461/files#r234416441

File: modules/test/test_issue.py (hunk 22 above)
Comment (amuritna) [reply modules/test/test_issue.py]: Ohh, okay! Thanks!

File: modules/test/test_issue.py [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+        self.input = MagicMock()
+        
+    @patch('modules.issue.post')
+    def test_success(self, mock_post):
+        mock_response = MagicMock()
+        mock_response.json.return_value = {
+            'html_url': 'https://github.com/test/test'
+        }
+        mock_post.return_value = mock_response
+        mock_post.assert_called_with('https://api.github.com/repos/test/test/issues')
```
Comment (sushain97) [inline modules/test/test_issue.py]: Can you also verify that the headers and body are correct?

File: modules/issue.py [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+issue.py - create a new GitHub issue
+author: amuritna
+"""
+from web import is_up, post
+from requests import HTTPError
+import json
+
+gh_uri = 'https://api.github.com'
+allowed_owners = ['apertium'] # makes checking for valid owner/repo combo faster
+default_desc = 'This issue was automatically made by begiak, Apertium\'s beloved IRC bot, by the order of someone on #apertium. A human is yet to update the description.'
```
Comment (sushain97) [inline modules/issue.py]: Can you specify who?

**Comment by sushain97** [comment]
Also, it looks like the build is failing?

**Comment by amuritna** [comment]
I've created a gist of the chat log [https://gist.github.com/amuritna/08923d2cd34b25b1ca007131b3230d9a](here). Is that sufficient, of would adding it to the readme or something be preferable?

**Comment by amuritna** [comment]
Ah, shame, I missed the deadline ... :/

**Comment by amuritna** [comment]
I've updated the tests and issue.py itself. I haven't investigated into why the build failed, however.

**Comment by sushain97** [comment]
There seems to be a syntax error in your test. I will take care of fixing the other issue in master and you'll have to update your branch.

**Comment by sushain97** [comment]
@amuritna you can run the tests locally! That way, you don't have to wait for CI every time.

**Comment by amuritna** [comment]
Yeah, okay :)

**Comment by amuritna** [comment]
Still no idea what's up with test_success's post(expected_values) not being called :/

**Comment by sushain97** [comment]
I think you might need to move the `assert` after `issue.issue` is actually called? Until then, the mock hasn't been called.

**Comment by amuritna** [comment]
Yeah, that'd make sense. Unfortunately I haven't managed to get my local nosetests3 to work yet (granted, I don't think I been "diving" into it as much as I usually do) so I haven't done any debugging since spamming commits right into GitHub yesterday.

Just got an idea how to solve my problem with my local setup though. If it doesn't work I'll try my luck once again with the Travis CI. A single commit and a possible revert commit wouldn't be as annoying and spammy, I hope.

**Comment by amuritna** [comment]
Ow :)

**Comment by amuritna** [comment]
Realized I don't need to `nosetests3` the whole repo and that `python -m unittest test_issue` works great. Testing/debugging locally now :)

**Comment by amuritna** [comment]
Build's a success, now! :)

File: modules/issue.py [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+    req_headers = {'Authorization': 'token {}'.format(oauth_token)}
+    req_body = json.dumps({
+        "title": title,
+        "body": default_desc.format(input.nick)
+    })
+
+    try:
+        req_str = post(req_url, req_body, req_headers)
+        req_json = json.loads(req_str)
+        return phenny.reply('Issue created. You can add a description at {}'.format(req_json['html_url']))
```
Comment (sushain97) [inline modules/issue.py]:
```suggestion
        phenny.reply('Issue created. You can add a description at {}'.format(req_json['html_url']))
```

File: modules/issue.py [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+    except HTTPError:
+        return phenny.reply('It appears that that repository does not exist.')
```
Comment (sushain97) [inline modules/issue.py]:
```suggestion
        phenny.reply('It appears that the repository you provided does not exist.')
```

**Comment by amuritna** [comment]
Yeah, I had already made the `.format()` changes (last thing I've been working on before I did the unit tests) when I produced [this chat log](https://gist.github.com/amuritna/08923d2cd34b25b1ca007131b3230d9a). And I've tested the unit tests locally.
